File: src-python/planarvibe/layouts/schnyder.py
"""Schnyder grid drawing. Port of static/js/layout-schnyder.js."""

from __future__ import annotations

import math
from typing import Sequence

from .. import geometry as geo
from .. import graph as gu
from .. import preprocessing

SCHNYDER_PREPARE_OPTIONS = {"triangulateOuterFace": True}


def _build_rotation_by_id(embedding: dict) -> dict[str, list[str]]:
    out: dict[str, list[str]] = {}
    rot = embedding["rotation"]
    for i, nid in enumerate(embedding["idByIndex"]):
        out[str(nid)] = [str(x) for x in (rot[i] if i < len(rot) else [])]
    return out


def _contract(node_ids, adjacency, a, b, c) -> list[str]:
    N = sum(len(adjacency.get(nid, [])) for nid in node_ids) // 2
    marked: dict[str, bool] = {nid: False for nid in node_ids}
    deg: dict[str, int] = {nid: 0 for nid in node_ids}
    L: list[str] = []
    candidates: list[str] = []

    marked[a] = marked[b] = marked[c] = True
    deg[a] = deg[b] = deg[c] = N

    an = list(adjacency.get(a, []))
    for x in an:
        marked[x] = True
        for y in adjacency.get(x, []):
            deg[y] = deg.get(y, 0) + 1
    for x in an:
        if deg[x] <= 2:
            candidates.append(x)

    while candidates:
        u = candidates.pop(0)
        if deg[u] != 2:
            continue
        L.insert(0, u)
        deg[u] = N
        for nb in adjacency.get(u, []):
            deg[nb] = deg.get(nb, 0) - 1
            if not marked.get(nb):
                marked[nb] = True
                for t in adjacency.get(nb, []):
                    deg[t] = deg.get(t, 0) + 1
                if deg[nb] <= 2:
                    candidates.append(nb)
            elif deg[nb] == 2:
                candidates.append(nb)
    return L


def _add_directed_labeled_edge(out_adj_by_label: dict, src: str, dst: str, label: int) -> None:
    a = out_adj_by_label[label]
    if src not in a:
        a[src] = []
    a[src].append(dst)


def _realizer(node_ids, L, a, b, c, rotation_by_id, adjacency) -> dict:
    ord_: dict[str, int] = {}
    i = 0
    ord_[b] = i; i += 1
    ord_[c] = i; i += 1
    for x in L:
        ord_[x] = i; i += 1
    ord_[a] = i; i += 1

    out_adj_by_label = {1: {}, 2: {}, 3: {}}
    for nid in node_ids:
        out_adj_by_label[1][nid] = []
        out_adj_by_label[2][nid] = []
        out_adj_by_label[3][nid] = []

    for v in L:
        rot = rotation_by_id.get(v, [])
        if not rot:
            return gu.build_layout_error({"reason": f"Missing rotation at vertex {v}"})
        first_idx = -1
        for j, w in enumerate(rot):
            if (ord_.get(w, 0)) > (ord_.get(v, 0)):
                first_idx = j
                break
        if first_idx < 0:
            return gu.build_layout_error({"reason": f"Could not find higher-order neighbor at vertex {v}"})

        idx1 = first_idx
        while (ord_.get(rot[idx1], 0)) > (ord_.get(v, 0)):
            idx1 = (idx1 + 1) % len(rot)
        _add_directed_labeled_edge(out_adj_by_label, rot[idx1], v, 2)

        idx2 = first_idx
        while (ord_.get(rot[idx2], 0)) > (ord_.get(v, 0)):
            idx2 = (idx2 - 1) % len(rot)
        _add_directed_labeled_edge(out_adj_by_label, rot[idx2], v, 3)

        walk = (idx1 + 1) % len(rot)
        while walk != idx2:
            _add_directed_labeled_edge(out_adj_by_label, v, rot[walk], 1)
            walk = (walk + 1) % len(rot)

    for x in adjacency.get(a, []):
        _add_directed_labeled_edge(out_adj_by_label, a, x, 1)
    _add_directed_labeled_edge(out_adj_by_label, b, a, 2)
    _add_directed_labeled_edge(out_adj_by_label, b, c, 2)
    _add_directed_labeled_edge(out_adj_by_label, c, a, 3)
    _add_directed_labeled_edge(out_adj_by_label, c, b, 3)

    return gu.build_layout_result({"ord": ord_, "outAdjByLabel": out_adj_by_label})


def _subtree_sizes(out_adj_by_label: dict, label: int, root: str) -> dict[str, int]:
    memo: dict[str, int] = {}
    visiting: dict[str, bool] = {}
    adj = out_adj_by_label[label]

    def dfs(v: str) -> int:
        if v in memo:
            return memo[v]
        if visiting.get(v):
            return 1
        visiting[v] = True
        s = 0
        for w in adj.get(v, []):
            s += dfs(w)
        visiting[v] = False
        memo[v] = s + 1
        return memo[v]

    import sys
    sys.setrecursionlimit(max(10000, sys.getrecursionlimit()))
    dfs(root)
    return memo


def _prefix_sum(out_adj_by_label: dict, label: int, root: str, val: dict[str, float]) -> dict[str, float]:
    summed: dict[str, float] = {}
    adj = out_adj_by_label[label]
    queue = [root]
    summed[root] = float(val.get(root, 0))
    head = 0
    while head < len(queue):
        v = queue[head]; head += 1
        for w in adj.get(v, []):
            if w in summed:
                continue
            summed[w] = float(val.get(w, 0)) + summed[v]
            queue.append(w)
    return summed


def _compute_schnyder_coordinates(node_ids, realizer_out, a, b, c) -> dict:
    out_adj = realizer_out["outAdjByLabel"]
    t1 = _subtree_sizes(out_adj, 1, a)
    t2 = _subtree_sizes(out_adj, 2, b)
    ones = {nid: 1.0 for nid in node_ids}
    p1 = _prefix_sum(out_adj, 1, a, ones)
    p2 = _prefix_sum(out_adj, 2, b, ones)
    p3 = _prefix_sum(out_adj, 3, c, ones)

    sum1 = _prefix_sum(out_adj, 2, b, {k: float(v) for k, v in t1.items()})
    sum1[a] = float(t1.get(a, 1))
    sum2 = _prefix_sum(out_adj, 3, c, {k: float(v) for k, v in t1.items()})
    sum2[a] = float(t1.get(a, 1))

    x: dict[str, float] = {}
    for v in node_ids:
        r1 = sum1.get(v, 0) + sum2.get(v, 0) - t1.get(v, 1)
        x[v] = r1 - p3.get(v, 1)

    sum1 = _prefix_sum(out_adj, 3, c, {k: float(v) for k, v in t2.items()})
    sum1[b] = float(t2.get(b, 1))
    sum2 = _prefix_sum(out_adj, 1, a, {k: float(v) for k, v in t2.items()})
    sum2[b] = float(t2.get(b, 1))

    y: dict[str, float] = {}
    for v in node_ids:
        r2 = sum1.get(v, 0) + sum2.get(v, 0) - t2.get(v, 1)
        y[v] = r2 - p1.get(v, 1)

    return {"x": x, "y": y}


def _build_screen_positions(coords: dict, node_ids: Sequence[str]) -> dict | None:
    min_x = math.inf
    max_y = -math.inf
    for nid in node_ids:
        xi = coords["x"].get(nid)
        yi = coords["y"].get(nid)
        if xi is None or yi is None or not math.isfinite(xi) or not math.isfinite(yi):
            return None
        if xi < min_x:
            min_x = xi
        if yi > max_y:
            max_y = yi
    if not math.isfinite(min_x) or not math.isfinite(max_y):
        return None
    SCALE = 30.0
    out: dict[str, tuple[float, float]] = {}
    for nid in node_ids:
        out[nid] = ((coords["x"][nid] - min_x) * SCALE + 20, (max_y - coords["y"][nid]) * SCALE + 20)
    return out


def _group_overlaps(pos_by_id: dict) -> list[list[str]]:
    buckets: dict[str, list[str]] = {}
    for nid, p in (pos_by_id or {}).items():
        if p is None or not math.isfinite(p[0]) or not math.isfinite(p[1]):
            continue
        k = f"{p[0]},{p[1]}"
        buckets.setdefault(k, []).append(nid)
    out: list[list[str]] = []
    for v in buckets.values():
        if len(v) > 1:
            v.sort()
            out.append(v)
    return out


def _count_overlap_extras(groups: list[list[str]]) -> int:
    return sum(len(g) - 1 for g in groups)


def _resolve_overlaps_without_crossings(pos_by_id: dict, edge_pairs: Sequence[tuple[str, str]]) -> dict | None:
    overlap_groups = _group_overlaps(pos_by_id)
    if not overlap_groups:
        return geo.copy_position_map(pos_by_id)
    pos = geo.copy_position_map(pos_by_id)
    DIRS = 24
    ring = [(math.cos(2 * math.pi * a / DIRS), math.sin(2 * math.pi * a / DIRS)) for a in range(DIRS)]
    radii = [0.25, 0.5, 0.75, 1.0, 1.5, 2.0, 3.0, 4.0, 6.0]

    for group in overlap_groups:
        anchor = pos[group[0]]
        placed = False
        for radius in radii:
            if placed:
                break
            for phase in range(len(ring)):
                if placed:
                    break
                trial = geo.copy_position_map(pos)
                for i, nid in enumerate(group):
                    idx = (phase + (i * len(ring)) // len(group)) % len(ring)
                    dx, dy = ring[idx]
                    trial[nid] = (anchor[0] + dx * radius, anchor[1] + dy * radius)
                if _group_overlaps(trial):
                    continue
                if geo.has_position_crossings(trial, edge_pairs):
                    continue
                pos = trial
                placed = True
        if not placed:
            return None
    if _group_overlaps(pos) or geo.has_position_crossings(pos, edge_pairs):
        return None
    return pos


def _candidate_outer_triples(emb: dict, rotation_by_id: dict) -> list[tuple[str, str, str]]:
    out: list[tuple[str, str, str]] = []
    edges = emb.get("edges") or []
    if not edges:
        return out
    e0 = edges[0]
    a = str(e0[0])
    b = str(e0[1])
    rot_b = rotation_by_id.get(b, [])
    try:
        idx_a = rot_b.index(a)
    except ValueError:
        return out
    c = str(rot_b[(idx_a - 1) % len(rot_b)])
    out.append((a, b, c))
    out.append((a, c, b))
    return out


def apply_layout(graph, initial_positions: dict | None = None, options: dict | None = None) -> dict:
    opts = options or {}
    prepared = preprocessing.prepare_graph_data(graph, {
        "failureLabel": "Schnyder",
        "currentPositions": initial_positions or {},
        "augmentationMethod": opts.get("augmentationMethod"),
        "augmentationOptions": SCHNYDER_PREPARE_OPTIONS,
    })
    if not prepared or not prepared.get("ok"):
        return gu.build_layout_error(prepared or {"message": "Schnyder failed"})

    emb = prepared["augmented"]["embedding"]
    rotation_by_id = _build_rotation_by_id(emb)
    augmented_graph = prepared["augmented"]["graph"]
    adjacency = {k: list(v) for k, v in augmented_graph.adjacency.items()}
    g = prepared["graph"]

    best_pos = None
    best_overlap_count = float("inf")
    candidates = _candidate_outer_triples(emb, rotation_by_id)
    if not candidates and emb.get("outerFace") and len(emb["outerFace"]) >= 3:
        of = emb["outerFace"]
        candidates = [(str(of[0]), str(of[1]), str(of[2]))]

    for (a, b, c) in candidates:
        L = _contract(augmented_graph.node_ids, adjacency, a, b, c)
        if len(L) != len(augmented_graph.node_ids) - 3:
            continue
        r = _realizer(augmented_graph.node_ids, L, a, b, c, rotation_by_id, adjacency)
        if not r.get("ok"):
            continue
        coords = _compute_schnyder_coordinates(augmented_graph.node_ids, r, a, b, c)
        pos = _build_screen_positions(coords, g.node_ids)
        if not pos:
            continue
        if geo.has_position_crossings(pos, g.edge_pairs):
            continue
        overlap_groups = _group_overlaps(pos)
        if overlap_groups:
            resolved = _resolve_overlaps_without_crossings(pos, g.edge_pairs)
            if resolved:
                pos = resolved
                overlap_groups = _group_overlaps(pos)
        overlap_count = _count_overlap_extras(overlap_groups)
        if overlap_count < best_overlap_count:
            best_overlap_count = overlap_count
            best_pos = pos
        if overlap_count == 0:
            break

    if best_pos is None:
        return gu.build_layout_error({"message": "Schnyder failed to find crossing-free embedding", "graph": g})

    final_pos = geo.normalize_position_map_to_viewport(best_pos)
    result = gu.build_layout_result({
        "nodeIds": g.node_ids,
        "edgePairs": g.edge_pairs,
        "graph": g,
        "positions": final_pos,
    })
    result["message"] = gu.build_layout_status_message("Schnyder layout", {"vertexCount": len(g.node_ids)})
    return result

File: src-cpp/src/schnyder.cpp
#include "layouts/schnyder.hpp"

#include "geometry.hpp"
#include "preprocessing.hpp"

#include <algorithm>
#include <cmath>
#include <deque>
#include <string>
#include <unordered_map>
#include <unordered_set>

namespace planarvibe::layouts {

namespace {

using StrVec = std::vector<std::string>;
using StrSet = std::unordered_set<std::string>;
using Adj = std::unordered_map<std::string, StrVec>;

Adj build_rotation_by_id(const planarity::StringEmbedding& emb) {
    Adj out;
    for (size_t i = 0; i < emb.id_by_index.size(); ++i) {
        out[emb.id_by_index[i]] = (i < emb.rotation.size()) ? emb.rotation[i] : StrVec{};
    }
    return out;
}

StrVec contract(const StrVec& node_ids, const Adj& adjacency,
                const std::string& a, const std::string& b, const std::string& c) {
    int N = 0;
    for (const auto& nid : node_ids) {
        auto it = adjacency.find(nid);
        if (it != adjacency.end()) N += (int)it->second.size();
    }
    N /= 2;
    std::unordered_map<std::string, bool> marked;
    std::unordered_map<std::string, int> deg;
    for (const auto& nid : node_ids) { marked[nid] = false; deg[nid] = 0; }
    marked[a] = marked[b] = marked[c] = true;
    deg[a] = deg[b] = deg[c] = N;

    auto ait = adjacency.find(a);
    const StrVec& an = (ait != adjacency.end()) ? ait->second : StrVec{};
    for (const auto& x : an) {
        marked[x] = true;
        auto xit = adjacency.find(x);
        if (xit == adjacency.end()) continue;
        for (const auto& y : xit->second) deg[y]++;
    }
    std::deque<std::string> candidates;
    for (const auto& x : an) if (deg[x] <= 2) candidates.push_back(x);

    StrVec L;
    while (!candidates.empty()) {
        std::string u = candidates.front();
        candidates.pop_front();
        if (deg[u] != 2) continue;
        L.insert(L.begin(), u);
        deg[u] = N;
        auto uit = adjacency.find(u);
        if (uit == adjacency.end()) continue;
        for (const auto& nb : uit->second) {
            deg[nb]--;
            if (!marked[nb]) {
                marked[nb] = true;
                auto nit = adjacency.find(nb);
                if (nit != adjacency.end()) {
                    for (const auto& t : nit->second) deg[t]++;
                }
                if (deg[nb] <= 2) candidates.push_back(nb);
            } else if (deg[nb] == 2) {
                candidates.push_back(nb);
            }
        }
    }
    return L;
}

struct RealizerOut {
    bool ok = false;
    std::string reason;
    std::unordered_map<std::string, int> ord;
    std::unordered_map<int, Adj> out_adj_by_label;  // label 1/2/3 -> src -> [dst]
};

void add_labeled_edge(RealizerOut& r, const std::string& src, const std::string& dst, int label) {
    r.out_adj_by_label[label][src].push_back(dst);
}

RealizerOut realizer(const StrVec& node_ids, const StrVec& L,
                     const std::string& a, const std::string& b, const std::string& c,
                     const Adj& rotation_by_id, const Adj& adjacency) {
    RealizerOut r;
    r.ord[b] = 0;
    r.ord[c] = 1;
    int i = 2;
    for (const auto& x : L) r.ord[x] = i++;
    r.ord[a] = i;

    for (int label : {1, 2, 3}) {
        r.out_adj_by_label[label] = {};
        for (const auto& nid : node_ids) r.out_adj_by_label[label][nid] = {};
    }

    auto ord_of = [&](const std::string& x) {
        auto it = r.ord.find(x);
        return (it != r.ord.end()) ? it->second : 0;
    };

    for (const auto& v : L) {
        auto rit = rotation_by_id.find(v);
        if (rit == rotation_by_id.end() || rit->second.empty()) {
            r.reason = "Missing rotation at vertex " + v;
            return r;
        }
        const auto& rot = rit->second;
        int first_idx = -1;
        for (int j = 0; j < (int)rot.size(); ++j) {
            if (ord_of(rot[j]) > ord_of(v)) { first_idx = j; break; }
        }
        if (first_idx < 0) {
            r.reason = "Could not find higher-order neighbor at vertex " + v;
            return r;
        }
        int idx1 = first_idx;
        while (ord_of(rot[idx1]) > ord_of(v)) idx1 = (idx1 + 1) % (int)rot.size();
        add_labeled_edge(r, rot[idx1], v, 2);

        int idx2 = first_idx;
        while (ord_of(rot[idx2]) > ord_of(v)) idx2 = (idx2 - 1 + (int)rot.size()) % (int)rot.size();
        add_labeled_edge(r, rot[idx2], v, 3);

        int walk = (idx1 + 1) % (int)rot.size();
        while (walk != idx2) {
            add_labeled_edge(r, v, rot[walk], 1);
            walk = (walk + 1) % (int)rot.size();
        }
    }

    auto ait = adjacency.find(a);
    if (ait != adjacency.end()) {
        for (const auto& x : ait->second) add_labeled_edge(r, a, x, 1);
    }
    add_labeled_edge(r, b, a, 2);
    add_labeled_edge(r, b, c, 2);
    add_labeled_edge(r, c, a, 3);
    add_labeled_edge(r, c, b, 3);
    r.ok = true;
    return r;
}

// Non-recursive subtree sizes. Mirrors the Python recursive DFS in which a
// recursive call on a currently-visiting vertex returns 1 (cycle guard).
// Iterative form: post-order walk; on-stack kids contribute 1 directly to
// their parent's sum rather than being pushed again.
std::unordered_map<std::string, int> subtree_sizes(const Adj& adj, const std::string& root) {
    std::unordered_map<std::string, int> memo;
    std::vector<std::pair<std::string, int>> stk;  // (node, child index)
    std::unordered_set<std::string> on_stack;
    stk.push_back({root, 0});
    on_stack.insert(root);
    while (!stk.empty()) {
        auto& [v, ci] = stk.back();
        auto it = adj.find(v);
        const StrVec& kids = (it != adj.end()) ? it->second : StrVec{};
        if (ci < (int)kids.size()) {
            const auto& w = kids[ci++];
            if (!memo.count(w) && !on_stack.count(w)) {
                stk.push_back({w, 0});
                on_stack.insert(w);
            }
            // If w is on_stack (cycle) or already memoized, we don't push;
            // we'll pick it up in the parent's sum below (memo[w] = 1 for
            // cycle kids since Python's dfs returns 1 for `visiting[v]`).
        } else {
            int s = 0;
            for (const auto& w : kids) {
                auto mit = memo.find(w);
                if (mit != memo.end()) s += mit->second;
                else s += 1;  // cycle guard: Python returns 1 for on-stack kids
            }
            memo[v] = s + 1;
            on_stack.erase(v);
            stk.pop_back();
        }
    }
    return memo;
}

std::unordered_map<std::string, double> prefix_sum(const Adj& adj, const std::string& root,
                                                    const std::unordered_map<std::string, double>& val) {
    std::unordered_map<std::string, double> summed;
    std::deque<std::string> queue{root};
    auto vroot = val.find(root);
    summed[root] = (vroot != val.end()) ? vroot->second : 0.0;
    while (!queue.empty()) {
        std::string v = queue.front();
        queue.pop_front();
        auto it = adj.find(v);
        if (it == adj.end()) continue;
        for (const auto& w : it->second) {
            if (summed.count(w)) continue;
            auto vit = val.find(w);
            summed[w] = (vit != val.end() ? vit->second : 0.0) + summed[v];
            queue.push_back(w);
        }
    }
    return summed;
}

struct SchnyderCoords {
    std::unordered_map<std::string, double> x, y;
};

SchnyderCoords compute_coordinates(const StrVec& node_ids, const RealizerOut& ro,
                                    const std::string& a, const std::string& b, const std::string& c) {
    const auto& L1 = ro.out_adj_by_label.at(1);
    const auto& L2 = ro.out_adj_by_label.at(2);
    const auto& L3 = ro.out_adj_by_label.at(3);
    auto t1 = subtree_sizes(L1, a);
    auto t2 = subtree_sizes(L2, b);
    std::unordered_map<std::string, double> ones;
    for (const auto& nid : node_ids) ones[nid] = 1.0;
    auto p1 = prefix_sum(L1, a, ones);
    (void)p1;
    auto p3 = prefix_sum(L3, c, ones);
    // x
    std::unordered_map<std::string, double> t1d;
    for (const auto& [k, v] : t1) t1d[k] = double(v);
    auto sum1 = prefix_sum(L2, b, t1d);
    sum1[a] = double(t1.count(a) ? t1.at(a) : 1);
    auto sum2 = prefix_sum(L3, c, t1d);
    sum2[a] = double(t1.count(a) ? t1.at(a) : 1);
    SchnyderCoords out;
    for (const auto& v : node_ids) {
        double r1 = (sum1.count(v) ? sum1.at(v) : 0) + (sum2.count(v) ? sum2.at(v) : 0) - (t1.count(v) ? t1.at(v) : 1);
        out.x[v] = r1 - (p3.count(v) ? p3.at(v) : 1);
    }
    // y
    std::unordered_map<std::string, double> t2d;
    for (const auto& [k, v] : t2) t2d[k] = double(v);
    sum1 = prefix_sum(L3, c, t2d);
    sum1[b] = double(t2.count(b) ? t2.at(b) : 1);
    sum2 = prefix_sum(L1, a, t2d);
    sum2[b] = double(t2.count(b) ? t2.at(b) : 1);
    auto p1b = prefix_sum(L1, a, ones);
    for (const auto& v : node_ids) {
        double r2 = (sum1.count(v) ? sum1.at(v) : 0) + (sum2.count(v) ? sum2.at(v) : 0) - (t2.count(v) ? t2.at(v) : 1);
        out.y[v] = r2 - (p1b.count(v) ? p1b.at(v) : 1);
    }
    return out;
}

std::optional<pg::PosByStr> build_screen_positions(const SchnyderCoords& c, const StrVec& node_ids) {
    double min_x = std::numeric_limits<double>::infinity(), max_y = -min_x;
    for (const auto& nid : node_ids) {
        auto xit = c.x.find(nid), yit = c.y.find(nid);
        if (xit == c.x.end() || yit == c.y.end()) return std::nullopt;
        if (!std::isfinite(xit->second) || !std::isfinite(yit->second)) return std::nullopt;
        if (xit->second < min_x) min_x = xit->second;
        if (yit->second > max_y) max_y = yit->second;
    }
    const double SCALE = 30.0;
    pg::PosByStr out;
    for (const auto& nid : node_ids) {
        out[nid] = {(c.x.at(nid) - min_x) * SCALE + 20, (max_y - c.y.at(nid)) * SCALE + 20};
    }
    return out;
}

std::vector<StrVec> group_overlaps(const pg::PosByStr& pos) {
    std::unordered_map<std::string, StrVec> buckets;
    for (const auto& [nid, p] : pos) {
        if (!std::isfinite(p[0]) || !std::isfinite(p[1])) continue;
        char buf[80];
        std::snprintf(buf, sizeof(buf), "%.17g,%.17g", p[0], p[1]);
        buckets[buf].push_back(nid);
    }
    std::vector<StrVec> out;
    for (auto& [_, v] : buckets) {
        if (v.size() > 1) {
            std::sort(v.begin(), v.end());
            out.push_back(std::move(v));
        }
    }
    return out;
}

// Port of Python _resolve_overlaps_without_crossings. Tries to move
// overlapping points onto a small ring around the anchor, picking the first
// configuration that has no remaining overlaps AND no edge crossings.
std::optional<pg::PosByStr> resolve_overlaps_without_crossings(
    const pg::PosByStr& pos,
    const std::vector<std::pair<std::string,std::string>>& edge_pairs) {
    auto groups = group_overlaps(pos);
    if (groups.empty()) return pos;
    pg::PosByStr cur = pos;
    const int DIRS = 24;
    std::vector<std::pair<double,double>> ring;
    ring.reserve(DIRS);
    for (int a = 0; a < DIRS; ++a) {
        double th = 2.0 * M_PI * a / DIRS;
        ring.emplace_back(std::cos(th), std::sin(th));
    }
    static const double radii[] = {0.25, 0.5, 0.75, 1.0, 1.5, 2.0, 3.0, 4.0, 6.0};

    // Helper: does trial have any crossings on edge_pairs?
    auto has_crossings = [&](const pg::PosByStr& trial) {
        // Build local int index over the trial's nodes.
        std::unordered_map<std::string,int> idx;
        for (const auto& [nid, _] : trial) { idx[nid] = (int)idx.size(); }
        PositionMap pm;
        pm.resize((int)idx.size());
        for (const auto& [nid, p] : trial) pm.put(idx[nid], p[0], p[1]);
        std::vector<std::pair<int,int>> eints;
        for (const auto& [u, v] : edge_pairs) {
            auto iu = idx.find(u), iv = idx.find(v);
            if (iu == idx.end() || iv == idx.end()) continue;
            eints.emplace_back(iu->second, iv->second);
        }
        return geo::has_position_crossings(pm, eints);
    };

    for (const auto& group : groups) {
        auto anchor = cur[group[0]];
        bool placed = false;
        for (double radius : radii) {
            if (placed) break;
            for (int phase = 0; phase < (int)ring.size() && !placed; ++phase) {
                pg::PosByStr trial = cur;
                for (size_t i = 0; i < group.size(); ++i) {
                    int idxr = (phase + (int)((i * ring.size()) / group.size())) % (int)ring.size();
                    trial[group[i]] = {anchor[0] + ring[idxr].first * radius,
                                        anchor[1] + ring[idxr].second * radius};
                }
                if (!group_overlaps(trial).empty()) continue;
                if (has_crossings(trial)) continue;
                cur = trial;
                placed = true;
            }
        }
        if (!placed) return std::nullopt;
    }
    if (!group_overlaps(cur).empty()) return std::nullopt;
    if (has_crossings(cur)) return std::nullopt;
    return cur;
}

int count_overlap_extras(const std::vector<StrVec>& groups) {
    int n = 0;
    for (const auto& g : groups) n += (int)g.size() - 1;
    return n;
}

std::vector<std::tuple<std::string, std::string, std::string>>
candidate_outer_triples(const planarity::StringEmbedding& emb, const Adj& rotation_by_id) {
    std::vector<std::tuple<std::string, std::string, std::string>> out;
    if (emb.edges.empty()) return out;
    const auto& e0 = emb.edges[0];
    std::string a = e0.first, b = e0.second;
    auto rit = rotation_by_id.find(b);
    if (rit == rotation_by_id.end()) return out;
    const auto& rot_b = rit->second;
    auto ait = std::find(rot_b.begin(), rot_b.end(), a);
    if (ait == rot_b.end()) return out;
    int idx_a = (int)(ait - rot_b.begin());
    std::string c = rot_b[(idx_a - 1 + (int)rot_b.size()) % rot_b.size()];
    out.emplace_back(a, b, c);
    out.emplace_back(a, c, b);
    return out;
}

} // namespace

LayoutResult schnyder(const Graph& g, const pg::PosByStr* initial_positions) {
    LayoutResult r;
    preprocessing::PrepareConfig cfg;
    cfg.failure_label = "Schnyder";
    cfg.current_positions = initial_positions;
    cfg.triangulate_outer_face = true;
    auto prep = preprocessing::prepare_graph_data(g, cfg);
    if (!prep.ok) {
        r.ok = false;
        r.message = prep.message.empty() ? "Schnyder failed" : prep.message;
        return r;
    }

    auto rotation_by_id = build_rotation_by_id(prep.augmented_embedding);
    Adj adjacency;
    for (const auto& nid : prep.augmented_node_ids) adjacency[nid] = {};
    for (const auto& [u, v] : prep.augmented_edge_pairs) {
        adjacency[u].push_back(v);
        adjacency[v].push_back(u);
    }

    pg::PosByStr best_pos;
    int best_overlap = std::numeric_limits<int>::max();
    bool have = false;
    auto candidates = candidate_outer_triples(prep.augmented_embedding, rotation_by_id);
    if (candidates.empty() && prep.augmented_embedding.outer_face.size() >= 3) {
        const auto& of = prep.augmented_embedding.outer_face;
        candidates.emplace_back(of[0], of[1], of[2]);
    }

    for (const auto& [a, b, c] : candidates) {
        auto L = contract(prep.augmented_node_ids, adjacency, a, b, c);
        if ((int)L.size() != (int)prep.augmented_node_ids.size() - 3) continue;
        auto ro = realizer(prep.augmented_node_ids, L, a, b, c, rotation_by_id, adjacency);
        if (!ro.ok) continue;
        auto coords = compute_coordinates(prep.augmented_node_ids, ro, a, b, c);
        auto screen = build_screen_positions(coords, g.node_names);
        if (!screen) continue;
        // Check crossings on the original graph only.
        PositionMap pm;
        pm.resize(g.n);
        for (int i = 0; i < g.n; ++i) {
            auto it = screen->find(g.node_names[i]);
            if (it != screen->end()) pm.put(i, it->second[0], it->second[1]);
        }
        if (geo::has_position_crossings(pm, g.edges)) continue;
        auto groups = group_overlaps(*screen);
        pg::PosByStr pos_use = *screen;
        if (!groups.empty()) {
            auto resolved = resolve_overlaps_without_crossings(*screen,
                [&]{ std::vector<std::pair<std::string,std::string>> eps;
                     for (const auto& [u, v] : g.edges) eps.emplace_back(g.node_names[u], g.node_names[v]);
                     return eps;
                }());
            if (resolved) {
                pos_use = *resolved;
                groups = group_overlaps(pos_use);
            }
        }
        int overlap = count_overlap_extras(groups);
        if (overlap < best_overlap) {
            best_overlap = overlap;
            best_pos = pos_use;
            have = true;
        }
        if (overlap == 0) break;
    }

    if (!have) {
        r.ok = false;
        r.message = "Schnyder failed to find crossing-free embedding";
        return r;
    }

    // Normalize to viewport.
    PositionMap pm;
    pm.resize((int)g.node_names.size());
    for (int i = 0; i < g.n; ++i) {
        auto it = best_pos.find(g.node_names[i]);
        if (it != best_pos.end()) pm.put(i, it->second[0], it->second[1]);
    }
    auto normalized = geo::normalize_position_map_to_viewport(pm);
    r.positions.resize(g.n);
    for (int i = 0; i < g.n; ++i) {
        if (normalized.has(i)) r.positions.put(i, normalized.pos[i][0], normalized.pos[i][1]);
    }
    r.ok = true;
    r.message = "Schnyder layout";
    return r;
}

} // namespace planarvibe::layouts
